File: src/pixiplex/javascript/widget.js
import * as pn from "./pixinet.js";

export const WIDGET_MESSAGES = {
  SYNC_MODEL: "msg:sync_model",
  SYNC_NODE_COORDINATES: "msg:sync_node_coordinates",
  CENTER: "msg:center",
};

const to_array = (value) => (Array.isArray(value) ? value : []);

const sync_coordinates = (model, pp) => {
  const nodes = Array.isArray(pp.nodes_gfx) ? pp.nodes_gfx : pp.nodes;
  model.set("_x", nodes.map((node) => Number(node.x) || 0));
  model.set("_y", nodes.map((node) => Number(node.y) || 0));
  model.save_changes();
};

const normalize_node_style = (style = null) => {
  if (!style || typeof style !== "object") return null;
  const lineStyle = style.lineStyle && typeof style.lineStyle === "object"
    ? {
        size: Number(style.lineStyle.size ?? style.lineStyle.width ?? 1.5),
        color: style.lineStyle.color ?? 0xffffff,
      }
    : { size: Number(style.line_size ?? 1.5), color: style.line_color ?? 0xffffff };

  return {
    radius: Number(style.radius ?? 6),
    color: style.color ?? 0x650a5a,
    alpha: Number(style.alpha ?? 1),
    lineStyle,
  };
};

const normalize_line_style = (style = null) => {
  if (!style || typeof style !== "object") return null;
  return {
    lineWidth: Number(style.lineWidth ?? style.line_width ?? 1),
    color: style.color ?? 0x000000,
    alpha: Number(style.alpha ?? 1),
  };
};

const apply_node_style = (pp, style) => {
  const normalized = normalize_node_style(style);
  if (!normalized) return;
  pp.node_style = normalized;
  if (Array.isArray(pp.nodes_gfx)) {
    pn.build_nodes(pp.nodes_gfx, pn.default_node_styles(pp.nodes_gfx, pp.node_style));
  }
};

const apply_line_style = (pp, style) => {
  const normalized = normalize_line_style(style);
  if (!normalized) return;
  pp.line_style = normalized;
  if (pp.links_gfx) {
    pn.build_links(pp.links, pp.links_gfx, pp.line_style);
  }
};

const sync_model = (model, pp) => {
  sync_coordinates(model, pp);
  model.set("nodes", to_array(pp.nodes));
  model.set("links", to_array(pp.links));
  model.save_changes();
};

async function initialize() {
  window.pn = pn;
}

async function render({ model, el }) {
  const nodes = to_array(model.get("nodes"));
  const links = to_array(model.get("links"));
  const pp = new pn.Pixiplex(
    nodes,
    links,
    model.get("width"),
    model.get("height"),
    model.get("scale"),
    model.get("forces") || {},
  );

  apply_node_style(pp, model.get("node_style"));
  apply_line_style(pp, model.get("line_style"));

  await pp.init_all();
  window.pp = pp;
  el.appendChild(pp.view);

  pp.sim?.on("end", () => sync_coordinates(model, pp));

  const apply_graph_data = (center = false) => {
    pp.set_graph_data(to_array(model.get("nodes")), to_array(model.get("links")), {
      center,
      drag: false,
    });
  };

  model.on("change:nodes", () => apply_graph_data(false));
  model.on("change:links", () => apply_graph_data(false));
  model.on("change:forces", () => pp.set_forces(model.get("forces") || {}));
  model.on("change:node_style", () => apply_node_style(pp, model.get("node_style")));
  model.on("change:line_style", () => apply_line_style(pp, model.get("line_style")));

  model.on("change:width", () => {
    pp.width = Number(model.get("width")) || pp.width;
    pp.view.style.width = `${pp.width}px`;
  });
  model.on("change:height", () => {
    pp.height = Number(model.get("height")) || pp.height;
    pp.view.style.height = `${pp.height}px`;
  });
  model.on("change:scale", () => {
    pp.scale = Number(model.get("scale")) || pp.scale;
  });

  model.on("msg:custom", (msg) => {
    if (msg.type === WIDGET_MESSAGES.SYNC_MODEL) {
      sync_model(model, pp);
    } else if (msg.type === WIDGET_MESSAGES.SYNC_NODE_COORDINATES) {
      sync_coordinates(model, pp);
    } else if (msg.type === WIDGET_MESSAGES.CENTER) {
      pp.center_graph(msg.fit, msg.x, msg.y);
    }
  });
}

export default { initialize, render };

File: src/pixiplex/python/widget.py
"""Anywidget-backed Pixinet widget host class."""

from __future__ import annotations

import json
import pathlib
from typing import Any, Optional

import traitlets
from numpy.typing import ArrayLike

from .forces import ForceConfig
from .styles import EdgeStyle, NodeStyle

try:
    import anywidget
except (
    ModuleNotFoundError
):  # pragma: no cover - fallback for non-widget test environments
    anywidget = None


class _FallbackAnyWidget(traitlets.HasTraits):
    def send(self, _msg: dict[str, Any]) -> None:
        return None


AnyWidgetBase = anywidget.AnyWidget if anywidget is not None else _FallbackAnyWidget

OUTPUT_DIR = pathlib.Path(__file__).resolve().parent.parent / "static"


def _normalize_nodes_links(
    nodes: Optional[list[dict[str, Any]]],
    links: Optional[list[dict[str, Any]]],
    node_ids: Optional[ArrayLike],
    edgelist: Optional[ArrayLike],
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    if nodes is not None or links is not None:
        return list(nodes or []), list(links or [])

    normalized_nodes = [
        {"id": int(node_id)}
        for node_id in (list(node_ids) if node_ids is not None else [])
    ]
    if edgelist is None:
        return normalized_nodes, []

    normalized_links = [
        {"source": int(edge[0]), "target": int(edge[1])} for edge in list(edgelist)
    ]
    return normalized_nodes, normalized_links


class Pixinet(AnyWidgetBase):
    """Python host for the Pixi/d3 force-directed graph widget."""

    _esm = OUTPUT_DIR / "widget.js"
    _css = OUTPUT_DIR / "widget.css"

    scale = traitlets.Float(2.0).tag(sync=True)
    width = traitlets.Int(250).tag(sync=True)
    height = traitlets.Int(250).tag(sync=True)

    nodes = traitlets.List(trait=traitlets.Dict()).tag(sync=True)
    links = traitlets.List(trait=traitlets.Dict()).tag(sync=True)

    _x = traitlets.List(trait=traitlets.Float()).tag(sync=True)
    _y = traitlets.List(trait=traitlets.Float()).tag(sync=True)

    node_style = traitlets.Instance(NodeStyle).tag(sync=True, to_json=NodeStyle.as_dict)
    line_style = traitlets.Instance(EdgeStyle).tag(sync=True, to_json=EdgeStyle.as_dict)
    forces = traitlets.Instance(ForceConfig).tag(sync=True, to_json=ForceConfig.as_dict)

    def __init__(
        self,
        node_ids: Optional[ArrayLike] = None,
        edgelist: Optional[ArrayLike] = None,
        *,
        nodes: Optional[list[dict[str, Any]]] = None,
        links: Optional[list[dict[str, Any]]] = None,
        width: int = 250,
        height: int = 250,
        scale: float = 2.0,
        forces: Optional[ForceConfig] = None,
        node_style: Optional[NodeStyle] = None,
        line_style: Optional[EdgeStyle] = None,
    ) -> None:
        super().__init__()
        normalized_nodes, normalized_links = _normalize_nodes_links(
            nodes=nodes, links=links, node_ids=node_ids, edgelist=edgelist
        )
        self.width = width
        self.height = height
        self.scale = scale
        self.nodes = normalized_nodes
        self.links = normalized_links
        self.forces = ForceConfig() if forces is None else forces
        self.node_style = NodeStyle() if node_style is None else node_style
        self.line_style = EdgeStyle() if line_style is None else line_style

    def center(
        self, fit_zoom: bool = False, x: float | None = None, y: float | None = None
    ):
        msg = {"type": "msg:center", "fit": fit_zoom}
        msg["x"] = float(x) if x is not None else (self.width * self.scale / 2)
        msg["y"] = float(y) if y is not None else (self.height * self.scale / 2)
        self.send(msg)

    @property
    def x(self):
        self.send({"type": "msg:sync_node_coordinates"})
        return self._x

    @property
    def y(self):
        self.send({"type": "msg:sync_node_coordinates"})
        return self._y

    @property
    def graph(self) -> dict[str, list[dict[str, Any]]]:
        return {"nodes": list(self.nodes), "links": list(self.links)}

    def embed_html(self, path: pathlib.Path):
        import ipywidgets.embed

        ipywidgets.embed.embed_minimal_html(
            str(path.resolve()), views=[self], drop_defaults=False
        )

    def embed_state(self, path: pathlib.Path = None):
        from ipywidgets.embed import embed_data

        return embed_data(self)

    def to_standalone_html(
        self,
        path: pathlib.Path | str,
        *,
        inline_js: bool = True,
        title: str = "Pixiplex Export",
    ) -> pathlib.Path:
        path = pathlib.Path(path)
        pixinet_js = (OUTPUT_DIR / "pixinet.js").read_text()
        graph_json = json.dumps(self.graph)
        forces_json = json.dumps(ForceConfig.as_dict(self.forces, widget=self))
        node_style_json = json.dumps(NodeStyle.as_dict(self.node_style, widget=self))
        line_style_json = json.dumps(EdgeStyle.as_dict(self.line_style, widget=self))

        if inline_js:
            module_block = f"""
<script type=\"module\">
{pixinet_js}
const graph = {graph_json};
const pp = new Pixiplex(graph.nodes, graph.links, {self.width}, {self.height}, {self.scale}, {forces_json});
pp.node_style = {node_style_json};
pp.line_style = {line_style_json};
await pp.init_all();
document.getElementById("pixiplex_container").appendChild(pp.view);
window.pp = pp;
</script>
"""
        else:
            module_block = f"""
<script type=\"module\">
import {{ Pixiplex }} from "./pixinet.js";
const graph = {graph_json};
const pp = new Pixiplex(graph.nodes, graph.links, {self.width}, {self.height}, {self.scale}, {forces_json});
pp.node_style = {node_style_json};
pp.line_style = {line_style_json};
await pp.init_all();
document.getElementById("pixiplex_container").appendChild(pp.view);
window.pp = pp;
</script>
"""

        html = f"""<!doctype html>
<html lang=\"en\">
<head>
  <meta charset=\"utf-8\" />
  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\" />
  <title>{title}</title>
  <style>
    html, body {{ margin: 0; padding: 0; }}
    #pixiplex_container {{ width: {self.width}px; height: {self.height}px; overflow: hidden; }}
  </style>
</head>
<body>
  <div id=\"pixiplex_container\"></div>
  {module_block}
</body>
</html>
"""
        path.write_text(html)
        return path

    def embed_raw(self, path: pathlib.Path | str):
        return self.to_standalone_html(path)
